flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -106,14 +106,24 @@ static bool isAbsent(const fir::ExtendedValue &exv) {
   return !fir::getBase(exv);
 }
 
-/// Process calls to Maxval, Minval intrinsic functions
+/// Placeholder function for SUM/PRODUCT intrinsics when genSum()
+/// or genProduct() call genExtremumValProdSum(). It is an internal
+/// compiler error (ICE) if this function gets called.
+static void genInvalidChar(Fortran::lower::FirOpBuilder &builder,
+                           mlir::Location loc, mlir::Value resultBox,
+                           mlir::Value arrayBox, mlir::Value maskBox,
+                           llvm::StringRef errMsg) {
+  fir::emitFatalError(loc, errMsg);
+}
+
+/// Process calls to Maxval, Minval, Product, Sum intrinsic functions
 template <typename FN, typename FD, typename FC>
 static fir::ExtendedValue
-genExtremumval(FN func, FD funcDim, FC funcChar, mlir::Type resultType,
-               Fortran::lower::FirOpBuilder &builder, mlir::Location loc,
-               Fortran::lower::StatementContext *stmtCtx,
-               llvm::StringRef errMsg,
-               llvm::ArrayRef<fir::ExtendedValue> args) {
+genExtremumValProdSum(FN func, FD funcDim, FC funcChar, mlir::Type resultType,
```

Comment:
I think I would rather make a new function for Product and Sum given characters do not make sense for PRODUCT/SUM and complex do not make sense for extremums. It would make things easier to read (at least for me).

You can avoid duplication by grouping the allocatable clean-up in a shared function (that could actually be shared by all functions inserting this clean-up, I do think box category check makes adds a lot, and we could find less repetitive ways to make it).

Reply:
I have broken these up. Also, refactored some common code.